src/data_structures.rs

```diff
@@ -25,23 +30,17 @@ pub struct VerifierKey<F: PrimeField, PC: PolynomialCommitment<F>>
     pub index_comms: Vec<PC::Commitment>,
 }
 
-impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::ToBytes
-    for VerifierKey<F, PC>
-{
-    fn write<W: std::io::Write>(&self, mut w: W) -> std::io::Result<()> {
-        self.index_info.write(&mut w)?;
-        self.index_comms.write(&mut w)
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::ToBytes for VerifierKey<F, PC> {
+    #[inline]
+    fn write<W: Write>(&self, writer: W) -> std::io::Result<()> {
+        self.index_comms.serialize_uncompressed(writer)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", e)))
     }
 }
 
-impl<F: PrimeField, PC: PolynomialCommitment<F>> Clone
-    for VerifierKey<F, PC>
-{
-    fn clone(&self) -> Self {
-        Self {
-            index_comms: self.index_comms.clone(),
-            index_info: self.index_info.clone(),
-        }
+impl<F: PrimeField, PC: PolynomialCommitment<F>> algebra::SemanticallyValid for VerifierKey<F, PC> {
+    fn is_valid(&self) -> bool {
```

Comment:
We could widen the notion of validity of a verifier key beyond just commitment checks, by some general checks on the index_info.